src/chunk.writer.cpp

```diff
@@ -190,14 +191,47 @@ ChunkWriter::write(const uint8_t* beg, const uint8_t* end)
     return bytes_out;
 }
 
+void
+ChunkWriter::write_frames_()
+{
+    const size_t bpt = bytes_per_tile(image_shape_, tile_shape_);
+    CHECK(bpt > 0);
+
+    if (buffer_.size() < bpt) {
+        buffer_.resize(bpt);
+    }
+
+    uint8_t* data = buffer_.data();
+
+    for (auto& frame : frames_) {
+        // Frames are inserted in order, but there may be gaps in the sequence
+        // because we could have received frames from other threads.
+        // If we encounter a gap, we need to stop writing.
+        if (frame->frame_id() != last_frame_ + std::pow(2, level_of_detail_)) {
+            break;
+        }
+        size_t nbytes = frame->copy_tile(data, bpt, tile_col_, tile_row_, tile_plane_);
+        nbytes = write(data, data + nbytes);
+
+        EXPECT(nbytes == bpt,
+               "Expected to write %lu bytes, but wrote %lu bytes.",
+               bpt,
+               nbytes);
+
+        last_frame_ = frame->frame_id();
+    }
+
+    frames_.clear();
```

Comment:
this is a problem. Need to remove the frames that got processed and leave the rest. Might be easiest to switch to a priority queue.